### ISSUE TITLE
Projen should allow setting project references for typescript projects

### ISSUE THREAD

**Issue by timoheinonen** [projen/projen#2434]
I have a monorepo project with multiple services (different aws cdk app projects) and a shared library for common constructs (own aws cdk construct library project). In future, I will probably move the aws cdk construct library to own repository and use it through npm registry, but for now, I would like to use it directly from the same repo. 

Typescript's project references (https://www.typescriptlang.org/docs/handbook/project-references.html) seems to be the best way to do this, but projen does't provide a way to set the tsconfig parameter "references" using TypescriptConfigOptions. Also, in the library project I would need to set tsconfig parameter "composite" to true, but the whole parameter is missing from TypescriptConfigOptions compilerOptions.

Would it be possible to add these to projen?

**Comment by mrgrain** [comment]
Yes please! PRs are welcome.

In the meantime, you can easily use overrides:
```ts
      project.tsconfig?.file.addOverride('compilerOptions.composite', true);
      project.tsconfig?.file.addOverride('references', []);
```

**Comment by skleinjung** [comment]
I have built this feature into a custom TypescriptProject subtype I use on my projects, but think it's something that could be submitted upstream pretty easily. However, there are two "flavors" of this and I'd like some feedback from the maintainers about which fits the project philosophy better.

1. The simplest and least "batteries included" version is to add a "references" property that takes a list of string, and directly sets it as the references.
2. A more advanced option (and what I use personally) is that "references" is a list of TypescriptProject references. The paths to  set in tsconfig are the relative paths between the referencing project's `outdir` and the referenced project's `outdir`. This requires all the projects to be projen-ified, but allows them to be synthed to any folder and the reference property doesn't need to care.
3. The most opinionated option, and what I will probably move to in my projects, is to not have a "references" property at all. It requires all projects (referencing and references) to share a common parent. Then, all the dependencies are checked against the parent's subprojects. If the dependency is a subproject of the common parent, a reference is added using the path logic in #2.

#3 supports the custom monorepo structure I have built around projen, and does what is to me the most sensible thing without custom properties. It might need to be expanded to include a full parent project hierarchy, although I personally don't nest projects more than one level deep.

If we can have a discussion about preferences on how this feature would be introduced, I should be able to find some time to work on it.

**Comment by skleinjung** [comment]
Yeah, okay. Some good points there about supporting both the fully opinionated approach while still allowing a fallback to explicitly set references manually. My understanding of the project's philosophy is that the most powerful "expected" behavior should be the default, so I agree that we should probably enable references by default.

I will look at putting together an API/documentation PR per your suggestion as a starting point and we can iterate on that.

**Comment by skleinjung** [comment]
Created a draft PR that shows what an API for fully generic support of references in TypescriptConfig could look like. My thoughts on how to proceed are:

- Confirm that these API signatures look good, and then implement/add tests/etc.
- Merge that generic support as its own, initial chunk of work
- Iterate on a proposed API for opinionated reference support as discussed above that uses the generic type as a baseline
- Implement and submit an implementation of that

Let me know if that sounds good, and I'll look for any comments on the draft before moving forward.
